### ISSUE TITLE
Periodically force GC to avoid a memory leak when running in the editor

### ISSUE THREAD

**Issue by kring** [CesiumGS/cesium-unreal#340]
We've noticed that memory usage rises continually inside the editor. It seems to be less of an issue (maybe not an issue at all) in shipping builds and even in Play-in-Editor mode.

I'm pasting an email conversation I had about this previously, just in case it's helpful. Note that at the time, Chaos was the default physics engine in the pre-release version of 4.26.

Kevin:
> Speaking of running a long time, though, I think we're currently leaking memory pretty badly. I spent some time today trying to understand why, and I'm even more confused. From what I can tell, the UCesiumGltfComponent instances that are created for each loaded tile are never being destroyed. This despite the fact that I'm explicitly calling DestroyComponent on them, and from the docs I've seen that's supposed to destroy the component even if other objects still reference it. But yet, neither the FinishDestroy method nor the component's destructor are called. Any idea what I might be missing? Take a look at destroyRecursively in ACesium3DTileset.cpp.
>
>  eventually figured out that I can ask the garbage collection system how an object is referenced:
>
```
Cmd: obj refs name=CesiumGltfComponent_208
LogReferenceChain: (root) GCObjectReferencer /Engine/Transient.GCObjectReferencer_0::AddReferencedObjects(): FChaosScene
LogReferenceChain:  (PendingKill) CesiumGltfPrimitiveComponent /Game/CesiumWorldTerrainAndOSMBuildings.CesiumWorldTerrainAndOSMBuildings:PersistentLevel.CesiumWorldTerrain.CesiumGltfComponent_208.CesiumGltfPrimitiveComponent_0->Outer
LogReferenceChain:   (PendingKill) CesiumGltfComponent /Game/CesiumWorldTerrainAndOSMBuildings.CesiumWorldTerrainAndOSMBuildings:PersistentLevel.CesiumWorldTerrain.CesiumGltfComponent_208
LogReferenceChain:   
```
>
> If I'm reading that correctly, something in Chaos is holding onto the CesiumGltfPrimitiveComponent (which is drived from UStaticMeshComponent). I haven't worked out why yet, though. Collision is already disabled for this tile (by SetCollisionEnabled(ECollisionEnabled::NoCollision)) by the time we get here.

Epic:
> I had some feedback related to memory. 
>
> In fact, the reference kept is something that the engine keeps when playing in editor. 
```
void FChaosScene::AddReferencedObjects(FReferenceCollector& Collector)
{
#if WITH_EDITOR
for(UObject* Obj : PieModifiedObjects)
{
Collector.AddReferencedObject(Obj);
}
#endif
}
```
> It should not happen when you run the packaged application. Did you notice this kind of memory growth when running a packaged app ? And as this member is protected, you'll not be able to do anything unless changing the engine, so you can keep it as is. 
> The normal way of suppressing a component is simply to remove the references to it. The engine will do the cleanup for you,  with the GC. (it should trigger at least one time per mn) 
> You can make an explicit call to suppress a component, but you still need to remove the reference ! 
```
myComponent->DestroyComponent();
myComponent = nullptr;
```

CC @baothientran

**Comment by baothientran** [comment]
@kring so I tried the command above and it seems like the Chaos system no longer holds the references to any of our component. 

```
Cmd: obj refs name=CesiumGltfComponent_88
LogReferenceChain: (PendingKill) CesiumGltfComponent /Game/CesiumWorldTerrainAndOSMBuildings.CesiumWorldTerrainAndOSMBuildings:PersistentLevel.Cesium3DTileset_4.CesiumGltfComponent_88 is not currently reachable
```

So it's pending kill, but never get collected by garbage collection. And after I force the garbage collection to do its job, I see the memory is reduced again

```
gc.CollectGarbageEveryFrame 1
```

**Comment by baothientran** [comment]
if I understand this engine code correctly https://github.com/EpicGames/UnrealEngine/blob/c3caf7b6bf12ae4c8e09b606f10a09776b4d1f38/Engine/Source/Runtime/Engine/Private/UnrealEngine.cpp#L1350, it looks like GC never call in the editor mode. It only get called when one of the console variables (the ones begin with `CVar....` and it can be triggered by the command above like `gc.CollectGarbageEveryFrame 1`) in the if statement are triggered or one of the worlds is in PIE mode. So it seems like we can't do anything about this

**Comment by baothientran** [comment]
Another way I can think of is to create an editor-only actor and use timer to trigger the garbage collection with 
`GEngine->ForceGarbageCollection(true)`

I'm not sure if it's a good practice, but at least we don't need to call it every frame

**Comment by baothientran** [comment]
so looks like providing in-editor-only actor to trigger GC every 1 or 2 second works quite good. The next step is to see if we can have any finer control on the garbage collection to prevent some small lags that happens when garbage collection is triggered

**Comment by jon-mullen** [comment]
I am having this issue or similar issue on Windows 10 with Unreal 4.26.2 (downloaded, not built) and Cesium plugin installed from marketplace.  It can be reproduced in the sample project by flying large distances in editor.  In standalone game the memory usage does come up very quickly but it eventually drops back down.

See screenshots for `stat memory` and `stat streaming` displays.

You all seem to have it under control but just fyi the leak is in the textures and it looks like you're counting on the GC to handle it when you remove from root in `UnrealResourcePreparer::freeRaster`.  You could try calling `UTexture::ConditionalBeginDestroy` on the texture here.

![stat_streaming](https://user-images.githubusercontent.com/10525877/125011307-352e1380-e036-11eb-87e0-36483f141c97.png)

![vlcsnap-2021-07-08-21h10m09s845](https://user-images.githubusercontent.com/10525877/125011312-395a3100-e036-11eb-8535-0e922630118b.png)

**Comment by kring** [comment]
This was largely addressed in #684.
